### PR TITLE
SONARJAVA-1366 - handling of primitive type cast

### PR DESCRIPTION
**Pull request by DidierBesset**

### CODE REVIEW

File: java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java [hunk 1]
```diff
@@ -287,6 +289,15 @@ private void visit(Tree tree, Tree terminator) {
           programState = ProgramState.put(programState, variableTree.symbol(), unstack.b.get(0));
         }
         break;
+      case TYPE_CAST:
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java]: MAJOR Reduce this switch case number of lines from 8 to at most 5, for example by extracting code into methods. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1151)

File: java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java [hunk 2]
```diff
@@ -287,6 +289,15 @@ private void visit(Tree tree, Tree terminator) {
...
+        TypeCastTree typeCast = (TypeCastTree) tree;
+        Type type = typeCast.type().symbolType();
+        if (type.isPrimitive()) {
+          Pair<ProgramState, List<SymbolicValue>> unstack = ProgramState.unstack(programState, 1);
+          programState = unstack.a;
+          programState = ProgramState.stackValue(programState, constraintManager.createSymbolicValue(typeCast.expression()));
+        }
+        break;
       case ASSIGNMENT:
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java]: MAJOR Reduce this switch case number of lines from 10 to at most 5, for example by extracting code into methods. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1151)

File: java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java [hunk 3]
```diff
@@ -287,6 +289,15 @@ private void visit(Tree tree, Tree terminator) {
...
         AssignmentExpressionTree assignmentExpressionTree = ((AssignmentExpressionTree) tree);
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java]: MAJOR Remove those useless parentheses. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AUselessParenthesesCheck)

File: java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java [hunk 4]
```diff
@@ -287,6 +289,15 @@ private void visit(Tree tree, Tree terminator) {
...
         // FIXME restricted to identifiers for now.
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1134)

**Comment by SonarTech** [comment]
SonarQube analysis reported 28 issues:
* BLOCKER 1 blocker
* MAJOR 26 major
* INFO 1 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* BLOCKER [NullPointerException might be thrown as 'terminator' is nullable here](https://github.com/SonarSource/sonar-java/blob/09b0c726da9076a154c7728975a28757558876d0/java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java#L194) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2259)
* MAJOR [Reduce this switch case number of lines from 6 to at most 5, for example by extracting code into methods.](https://github.com/SonarSource/sonar-java/blob/09b0c726da9076a154c7728975a28757558876d0/java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java#L326) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1151)
* MAJOR [The Cyclomatic Complexity of this method "handleBlockExit" is 16 which is greater than 10 authorized.](https://github.com/SonarSource/sonar-java/blob/09b0c726da9076a154c7728975a28757558876d0/java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java#L191) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
* MAJOR [Reduce this switch case number of lines from 7 to at most 5, for example by extracting code into methods.](https://github.com/SonarSource/sonar-java/blob/09b0c726da9076a154c7728975a28757558876d0/java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java#L349) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1151)
* MAJOR [Reduce this switch case number of lines from 7 to at most 5, for example by extracting code into methods.](https://github.com/SonarSource/sonar-java/blob/09b0c726da9076a154c7728975a28757558876d0/java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java#L365) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1151)
* MAJOR [This method has 163 lines, which is greater than the 100 lines authorized. Split it into smaller methods.](https://github.com/SonarSource/sonar-java/blob/09b0c726da9076a154c7728975a28757558876d0/java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java#L251) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS138)
* MAJOR [Rename this local variable name to match the regular expression '^[a-z][a-zA-Z0-9]*$'.](https://github.com/SonarSource/sonar-java/blob/09b0c726da9076a154c7728975a28757558876d0/java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java#L388) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS00117)
* MAJOR [Make programPosition a static final constant or non-public and provide accessors if needed.](https://github.com/SonarSource/sonar-java/blob/09b0c726da9076a154c7728975a28757558876d0/java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java#L86) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AClassVariableVisibilityCheck)
* MAJOR [Make "debugPrint" a "static" method.](https://github.com/SonarSource/sonar-java/blob/09b0c726da9076a154c7728975a28757558876d0/java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java#L465) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2325)
* MAJOR [Reduce this switch case number of lines from 8 to at most 5, for example by extracting code into methods.](https://github.com/SonarSource/sonar-java/blob/09b0c726da9076a154c7728975a28757558876d0/java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java#L382) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1151)
* ... 14 more

File: java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java [hunk 5]
```diff
@@ -287,6 +289,15 @@ private void visit(Tree tree, Tree terminator) {
           programState = ProgramState.put(programState, variableTree.symbol(), unstack.b.get(0));
         }
         break;
+      case TYPE_CAST:
+        TypeCastTree typeCast = (TypeCastTree) tree;
+        Type type = typeCast.type().symbolType();
+        if (type.isPrimitive()) {
+          Pair<ProgramState, List<SymbolicValue>> unstack = ProgramState.unstack(programState, 1);
+          programState = unstack.a;
+          programState = ProgramState.stackValue(programState, constraintManager.createSymbolicValue(typeCast.expression()));
```
Comment (benzonico) [inline java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java]: I tend to think that we shoud create the new symbolic value based on the type cast tree and not the expression.

**Pull request merged**